tiledb/common/thread_pool.cc

```diff
@@ -39,17 +40,45 @@ namespace tiledb {
 namespace common {
 
 // Define the static ThreadPool member variables.
-std::unordered_map<std::thread::id, ThreadPool*> ThreadPool::tp_index_;
-std::mutex ThreadPool::tp_index_lock_;
-std::unordered_map<std::thread::id, tdb_shared_ptr<ThreadPool::PackagedTask>>
-    ThreadPool::task_index_;
-std::mutex ThreadPool::task_index_lock_;
+template <>
+ThreadPool::tp_index_singleton_type::central_type
+    ThreadPool::tp_index_singleton_type::singleton_central;
+template <>
+std::mutex ThreadPool::tp_index_singleton_type::lock_central;
+
+template <>
+ThreadPool::task_index_singleton_type::central_type
+    ThreadPool::task_index_singleton_type::singleton_central;
+template <>
+std::mutex ThreadPool::task_index_singleton_type::lock_central;
+
+template <class T>
+typename class_singleton<T>::member_type class_singleton<T>::member_factory() {
+  std::lock_guard<std::mutex> lock(lock_central);
+  if (singleton_central.use_count() != 0) {
+    return singleton_central.lock();
+  }
+  /* Assert: singleton_central has a null stored pointer */
+  auto the_singleton = std::make_shared<singleton_type>();
+  if (!the_singleton) {
+    /* Assert: the_singleton has a null stored pointer */
```

Comment:
use an `assert()` for this

Reply:
We can, but these comments are mostly for tracking correctness proof in the flow of control. I avoid an explicit assert() if I've just checked it in code. Fir there's a precondition I want to rely on, that's when I'll put in an assert(), because there no guarantee the caller has satisfied the precondition.